File: MovieRandomizer.java
// MovieRandomizer.java
import java.io.*;
import java.nio.file.*;
import java.time.LocalDate;
import java.util.*;
import java.util.stream.Collectors;

record Movie(int id, String title, String genre, String director, int year, boolean watched, int rating, String notes, String addedDate) implements Serializable {}

class MoviesData implements Serializable {
    private static final long serialVersionUID = 1L;
    public List<Movie> movies;
}

class MovieRandomizer implements Serializable {
    private static final long serialVersionUID = 1L;
    private List<Movie> movies = new ArrayList<>();
    private int nextId = 1;

    public Movie addMovie(String title, String genre, String director, int year, boolean watched, int rating, String notes) {
        if (rating < 1 || rating > 10) throw new IllegalArgumentException("Оценка должна быть от 1 до 10");
        if (year < 1900 || year > LocalDate.now().getYear())
            throw new IllegalArgumentException("Год должен быть от 1900 до " + LocalDate.now().getYear());
        if (title.isBlank() || genre.isBlank())
            throw new IllegalArgumentException("Название и жанр не могут быть пустыми");
        if (director.isBlank()) director = "Неизвестен";
        if (notes == null) notes = "";
        Movie movie = new Movie(nextId, title, genre, director, year, watched, rating, notes, LocalDate.now().toString());
        movies.add(movie);
        nextId++;
        return movie;
    }

    public Optional<Movie> findMovie(int id) {
        return movies.stream().filter(m -> m.id() == id).findFirst();
    }

    public boolean editMovie(int id, Map<String, Object> updates) {
        Optional<Movie> opt = findMovie(id);
        if (opt.isEmpty()) return false;
        Movie old = opt.get();
        movies.remove(old);
        String title = (String) updates.getOrDefault("title", old.title());
        String genre = (String) updates.getOrDefault("genre", old.genre());
        String director = (String) updates.getOrDefault("director", old.director());
        int year = (int) updates.getOrDefault("year", old.year());
        boolean watched = (boolean) updates.getOrDefault("watched", old.watched());
        int rating = (int) updates.getOrDefault("rating", old.rating());
        String notes = (String) updates.getOrDefault("notes", old.notes());
        Movie updated = new Movie(old.id(), title, genre, director, year, watched, rating, notes, old.addedDate());
        movies.add(updated);
        return true;
    }

    public boolean deleteMovie(int id) {
        return movies.removeIf(m -> m.id() == id);
    }

    public List<Movie> getUnwatched() {
        return movies.stream().filter(m -> !m.watched()).collect(Collectors.toList());
    }

    public List<Movie> getByGenre(String genre) {
        return movies.stream().filter(m -> m.genre().equalsIgnoreCase(genre)).collect(Collectors.toList());
    }

    public Movie randomMovie(String genre) {
        List<Movie> pool = genre != null ? getByGenre(genre) : getUnwatched();
        if (pool.isEmpty()) {
            // если нет, берём все (или по жанру)
            pool = genre != null ? getByGenre(genre) : movies;
        }
        if (pool.isEmpty()) return null;
        Random rand = new Random();
        return pool.get(rand.nextInt(pool.size()));
    }

    public Map<String, Object> getStats() {
        int total = movies.size();
        int watchedCount = (int) movies.stream().filter(Movie::watched).count();
        int unwatched = total - watchedCount;
        double avgRating = movies.stream().filter(Movie::watched).mapToInt(Movie::rating).average().orElse(0);
        Map<String, Integer> genres = new HashMap<>();
        movies.forEach(m -> genres.put(m.genre(), genres.getOrDefault(m.genre(), 0) + 1));
        Map<String, Object> stats = new HashMap<>();
        stats.put("total", total);
        stats.put("watched", watchedCount);
        stats.put("unwatched", unwatched);
        stats.put("avg_rating", avgRating);
        stats.put("genres", genres);
        return stats;
    }

    public void saveToFile(String filename) throws IOException {
        MoviesData data = new MoviesData();
        data.movies = new ArrayList<>(movies);
        try (ObjectOutputStream oos = new ObjectOutputStream(Files.newOutputStream(Paths.get(filename)))) {
            oos.writeObject(data);
        }
    }

    public void loadFromFile(String filename) throws IOException, ClassNotFoundException {
        try (ObjectInputStream ois = new ObjectInputStream(Files.newInputStream(Paths.get(filename)))) {
            MoviesData data = (MoviesData) ois.readObject();
            movies = new ArrayList<>(data.movies);
            for (Movie m : movies) {
                if (m.id() >= nextId) nextId = m.id() + 1;
            }
        }
    }

    public void exportCSV(String filename) throws IOException {
        try (PrintWriter pw = new PrintWriter(Files.newBufferedWriter(Paths.get(filename)))) {
            pw.println("ID;Название;Жанр;Режиссёр;Год;Просмотрен;Оценка;Заметки;Дата добавления");
            for (Movie m : movies) {
                pw.printf("%d;%s;%s;%s;%d;%s;%d;%s;%s%n",
                        m.id(), m.title(), m.genre(), m.director(), m.year(),
                        m.watched() ? "Да" : "Нет", m.rating(), m.notes(), m.addedDate());
            }
        }
    }

    public void importCSV(String filename) throws IOException {
        try (BufferedReader br = Files.newBufferedReader(Paths.get(filename))) {
            String line = br.readLine(); // header
            while ((line = br.readLine()) != null) {
                String[] parts = line.split(";");
                if (parts.length < 9) continue;
                String title = parts[1];
                String genre = parts[2];
                String director = parts[3];
                int year = Integer.parseInt(parts[4]);
                boolean watched = parts[5].equals("Да");
                int rating = Integer.parseInt(parts[6]);
                String notes = parts[7];
                try {
                    addMovie(title, genre, director, year, watched, rating, notes);
                } catch (Exception e) {
                    System.out.println("Ошибка импорта строки: " + e.getMessage());
                }
            }
        }
    }

    public List<Movie> getMovies() { return Collections.unmodifiableList(movies); }
}

public class MovieRandomizerApp {
    private static final Scanner scanner = new Scanner(System.in);
    private static final Random rand = new Random();

    private static String readString(String prompt) {
        System.out.print(prompt);
        return scanner.nextLine().trim();
    }

    private static int readInt(String prompt) {
        while (true) {
            try {
                System.out.print(prompt);
                return Integer.parseInt(scanner.nextLine().trim());
            } catch (NumberFormatException e) {
                System.out.println("Введите число.");
            }
        }
    }

    private static boolean readBool(String prompt) {
        while (true) {
            String input = readString(prompt);
            if (input.equals("1")) return true;
            if (input.equals("0")) return false;
            System.out.println("Введите 1 или 0.");
        }
    }

    private static void printMovie(Movie movie) {
        String status = movie.watched() ? "✅ Просмотрен" : "⏳ Не просмотрен";
        System.out.printf("#%d - %s (%d)%n", movie.id(), movie.title(), movie.year());
        System.out.printf("   Жанр: %s, Режиссёр: %s%n", movie.genre(), movie.director());
        System.out.printf("   %s, Оценка: %d/10%n", status, movie.rating());
        if (!movie.notes().isBlank()) {
            System.out.printf("   Заметки: %s%n", movie.notes());
        }
        System.out.printf("   Добавлен: %s%n", movie.addedDate());
    }

    public static void main(String[] args) {
        MovieRandomizer randomizer = new MovieRandomizer();
        try {
            randomizer.loadFromFile("movies_data.ser");
        } catch (IOException | ClassNotFoundException e) {
            System.out.println("Не удалось загрузить данные.");
        }

        while (true) {
            System.out.println("\n===== ГЕНЕРАТОР СЛУЧАЙНОГО ФИЛЬМА (Java) =====");
            System.out.println("1. Добавить фильм");
            System.out.println("2. Показать все фильмы");
            System.out.println("3. Рекомендовать случайный фильм");
            System.out.println("4. Рекомендовать по жанру");
            System.out.println("5. Показать непросмотренные фильмы");
            System.out.println("6. Отметить фильм как просмотренный");
            System.out.println("7. Редактировать фильм");
            System.out.println("8. Удалить фильм");
            System.out.println("9. Показать статистику");
            System.out.println("10. Сохранить в файл");
            System.out.println("11. Загрузить из файла");
            System.out.println("12. Экспорт в CSV");
            System.out.println("13. Импорт из CSV");
            System.out.println("0. Выход");
            String choice = readString("Выберите действие: ");

            switch (choice) {
                case "0" -> { return; }
                case "1" -> {
                    String title = readString("Название: ");
                    if (title.isBlank()) { System.out.println("Название не может быть пустым."); continue; }
                    String genre = readString("Жанр: ");
                    if (genre.isBlank()) { System.out.println("Жанр не может быть пустым."); continue; }
                    String director = readString("Режиссёр (необязательно): ");
                    int year = readInt("Год выпуска: ");
                    boolean watched = readBool("Статус (1-просмотрен, 0-нет): ");
                    int rating = readInt("Оценка (1-10): ");
                    String notes = readString("Заметки (необязательно): ");
                    try {
                        Movie movie = randomizer.addMovie(title, genre, director, year, watched, rating, notes);
                        System.out.println("Фильм добавлен с ID " + movie.id());
                    } catch (Exception e) {
                        System.out.println("Ошибка: " + e.getMessage());
                    }
                }
                case "2" -> {
                    if (randomizer.getMovies().isEmpty()) System.out.println("Нет фильмов.");
                    else randomizer.getMovies().forEach(MovieRandomizerApp::printMovie);
                }
                case "3" -> {
                    Movie movie = randomizer.randomMovie(null);
                    if (movie == null) System.out.println("Нет фильмов в коллекции.");
                    else { System.out.println("\n🎬 Рекомендую посмотреть:"); printMovie(movie); }
                }
                case "4" -> {
                    String genre = readString("Введите жанр: ");
                    if (genre.isBlank()) { System.out.println("Введите жанр."); continue; }
                    Movie movie = randomizer.randomMovie(genre);
                    if (movie == null) System.out.println("Нет фильмов в жанре '" + genre + "'.");
                    else { System.out.println("\n🎬 Рекомендую посмотреть в жанре " + genre + ":"); printMovie(movie); }
                }
                case "5" -> {
                    var unwatched = randomizer.getUnwatched();
                    if (unwatched.isEmpty()) System.out.println("Нет непросмотренных фильмов.");
                    else unwatched.forEach(MovieRandomizerApp::printMovie);
                }
                case "6" -> {
                    int id = readInt("Введите ID фильма: ");
                    var opt = randomizer.findMovie(id);
                    if (opt.isEmpty()) { System.out.println("Фильм не найден."); continue; }
                    Movie old = opt.get();
                    if (old.watched()) System.out.println("Фильм уже отмечен как просмотренный.");
                    else {
                        randomizer.editMovie(id, Map.of("watched", true));
                        System.out.println("Фильм отмечен как просмотренный.");
                    }
                }
                case "7" -> {
                    int id = readInt("Введите ID фильма для редактирования: ");
                    var opt = randomizer.findMovie(id);
                    if (opt.isEmpty()) { System.out.println("Фильм не найден."); continue; }
                    Movie old = opt.get();
                    System.out.println("Оставьте поле пустым, чтобы не менять.");
                    String newTitle = readString("Название (" + old.title() + "): ");
                    String newGenre = readString("Жанр (" + old.genre() + "): ");
                    String newDirector = readString("Режиссёр (" + old.director() + "): ");
                    String newYearStr = readString("Год (" + old.year() + "): ");
                    String newWatchedStr = readString("Статус (1-просмотрен, 0-нет) сейчас: " + (old.watched() ? "1" : "0") + ": ");
                    String newRatingStr = readString("Оценка (" + old.rating() + "): ");
                    String newNotes = readString("Заметки (" + old.notes() + "): ");
                    Map<String, Object> updates = new HashMap<>();
                    if (!newTitle.isBlank()) updates.put("title", newTitle);
                    if (!newGenre.isBlank()) updates.put("genre", newGenre);
                    if (!newDirector.isBlank()) updates.put("director", newDirector);
                    if (!newYearStr.isBlank()) {
                        try { updates.put("year", Integer.parseInt(newYearStr)); }
                        catch (NumberFormatException e) { System.out.println("Год должен быть числом, пропускаем."); }
                    }
                    if (!newWatchedStr.isBlank()) updates.put("watched", newWatchedStr.equals("1"));
                    if (!newRatingStr.isBlank()) {
                        try { updates.put("rating", Integer.parseInt(newRatingStr)); }
                        catch (NumberFormatException e) { System.out.println("Оценка должна быть числом, пропускаем."); }
                    }
                    if (!newNotes.isBlank()) updates.put("notes", newNotes);
                    if (randomizer.editMovie(id, updates)) System.out.println("Фильм обновлён.");
                    else System.out.println("Ошибка обновления.");
                }
                case "8" -> {
                    int id = readInt("Введите ID фильма для удаления: ");
                    if (randomizer.deleteMovie(id)) System.out.println("Фильм удалён.");
                    else System.out.println("Фильм не найден.");
                }
                case "9" -> {
                    var stats = randomizer.getStats();
                    System.out.println("\n=== СТАТИСТИКА ===");
                    System.out.println("Всего фильмов: " + stats.get("total"));
                    System.out.println("Просмотрено: " + stats.get("watched"));
                    System.out.println("Не просмотрено: " + stats.get("unwatched"));
                    System.out.printf("Средняя оценка (просмотренные): %.2f%n", stats.get("avg_rating"));
                    System.out.println("По жанрам:");
                    @SuppressWarnings("unchecked")
                    Map<String, Integer> genres = (Map<String, Integer>) stats.get("genres");
                    genres.forEach((g, c) -> System.out.println("  " + g + ": " + c));
                }
                case "10" -> {
                    try {
                        randomizer.saveToFile("movies_data.ser");
                        System.out.println("Сохранено.");
                    } catch (IOException e) {
                        System.out.println("Ошибка сохранения: " + e.getMessage());
                    }
                }
                case "11" -> {
                    try {
                        randomizer.loadFromFile("movies_data.ser");
                        System.out.println("Загружено.");
                    } catch (IOException | ClassNotFoundException e) {
                        System.out.println("Ошибка загрузки: " + e.getMessage());
                    }
                }
                case "12" -> {
                    try {
                        randomizer.exportCSV("movies_export.csv");
                        System.out.println("Экспортировано в movies_export.csv");
                    } catch (IOException e) {
                        System.out.println("Ошибка экспорта: " + e.getMessage());
                    }
                }
                case "13" -> {
                    try {
                        randomizer.importCSV("movies_export.csv");
                        System.out.println("Импортировано из movies_export.csv");
                    } catch (IOException e) {
                        System.out.println("Ошибка импорта: " + e.getMessage());
                    }
                }
                default -> System.out.println("Неизвестная команда.");
            }
        }
    }
}

File: MovieRandomizer.cs
// MovieRandomizer.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;
using System.Text.Json.Serialization;

public record Movie(
    int Id,
    string Title,
    string Genre,
    string Director,
    int Year,
    bool Watched,
    int Rating,
    string Notes,
    string AddedDate
);

public class MoviesData
{
    public List<Movie> Movies { get; set; } = new();
}

public class MovieRandomizer
{
    private List<Movie> movies = new();
    private int nextId = 1;
    private static readonly Random rand = new();

    public IReadOnlyList<Movie> Movies => movies.AsReadOnly();

    public Movie AddMovie(string title, string genre, string director, int year, bool watched, int rating, string notes)
    {
        if (rating < 1 || rating > 10) throw new ArgumentException("Оценка должна быть от 1 до 10");
        if (year < 1900 || year > DateTime.Now.Year)
            throw new ArgumentException($"Год должен быть от 1900 до {DateTime.Now.Year}");
        if (string.IsNullOrWhiteSpace(title) || string.IsNullOrWhiteSpace(genre))
            throw new ArgumentException("Название и жанр не могут быть пустыми");
        if (string.IsNullOrWhiteSpace(director)) director = "Неизвестен";
        if (notes == null) notes = "";
        var movie = new Movie(nextId, title, genre, director, year, watched, rating, notes, DateTime.Now.ToString("yyyy-MM-dd"));
        movies.Add(movie);
        nextId++;
        return movie;
    }

    public Movie? FindMovie(int id) => movies.FirstOrDefault(m => m.Id == id);

    public bool EditMovie(int id, Dictionary<string, object> updates)
    {
        var old = FindMovie(id);
        if (old == null) return false;
        movies.Remove(old);
        string title = updates.ContainsKey("title") ? (string)updates["title"] : old.Title;
        string genre = updates.ContainsKey("genre") ? (string)updates["genre"] : old.Genre;
        string director = updates.ContainsKey("director") ? (string)updates["director"] : old.Director;
        int year = updates.ContainsKey("year") ? (int)updates["year"] : old.Year;
        bool watched = updates.ContainsKey("watched") ? (bool)updates["watched"] : old.Watched;
        int rating = updates.ContainsKey("rating") ? (int)updates["rating"] : old.Rating;
        string notes = updates.ContainsKey("notes") ? (string)updates["notes"] : old.Notes;
        var updated = new Movie(old.Id, title, genre, director, year, watched, rating, notes, old.AddedDate);
        movies.Add(updated);
        return true;
    }

    public bool DeleteMovie(int id) => movies.RemoveAll(m => m.Id == id) > 0;

    public List<Movie> GetUnwatched() => movies.Where(m => !m.Watched).ToList();

    public List<Movie> GetByGenre(string genre) =>
        movies.Where(m => string.Equals(m.Genre, genre, StringComparison.OrdinalIgnoreCase)).ToList();

    public Movie? RandomMovie(string genre = null)
    {
        var pool = string.IsNullOrEmpty(genre) ? GetUnwatched() : GetByGenre(genre);
        if (pool.Count == 0)
        {
            // если нет, берём все (или по жанру)
            pool = string.IsNullOrEmpty(genre) ? movies : GetByGenre(genre);
        }
        if (pool.Count == 0) return null;
        return pool[rand.Next(pool.Count)];
    }

    public Dictionary<string, object> GetStats()
    {
        int total = movies.Count;
        int watchedCount = movies.Count(m => m.Watched);
        int unwatched = total - watchedCount;
        double avgRating = movies.Where(m => m.Watched).Any() ? movies.Where(m => m.Watched).Average(m => m.Rating) : 0;
        var genres = movies.GroupBy(m => m.Genre).ToDictionary(g => g.Key, g => g.Count());
        return new Dictionary<string, object>
        {
            ["total"] = total,
            ["watched"] = watchedCount,
            ["unwatched"] = unwatched,
            ["avg_rating"] = avgRating,
            ["genres"] = genres
        };
    }

    public void SaveToFile(string filename)
    {
        var data = new MoviesData { Movies = movies };
        var options = new JsonSerializerOptions { WriteIndented = true };
        string json = JsonSerializer.Serialize(data, options);
        File.WriteAllText(filename, json);
    }

    public void LoadFromFile(string filename)
    {
        if (!File.Exists(filename)) return;
        string json = File.ReadAllText(filename);
        var data = JsonSerializer.Deserialize<MoviesData>(json);
        if (data != null)
        {
            movies = data.Movies;
            nextId = movies.Any() ? movies.Max(m => m.Id) + 1 : 1;
        }
    }

    public void ExportCSV(string filename)
    {
        using var writer = new StreamWriter(filename);
        writer.WriteLine("ID;Название;Жанр;Режиссёр;Год;Просмотрен;Оценка;Заметки;Дата добавления");
        foreach (var m in movies)
        {
            writer.WriteLine($"{m.Id};{m.Title};{m.Genre};{m.Director};{m.Year};{(m.Watched ? "Да" : "Нет")};{m.Rating};{m.Notes};{m.AddedDate}");
        }
    }

    public void ImportCSV(string filename)
    {
        if (!File.Exists(filename)) throw new FileNotFoundException("Файл не найден");
        using var reader = new StreamReader(filename);
        string header = reader.ReadLine(); // skip header
        while (!reader.EndOfStream)
        {
            string line = reader.ReadLine();
            var parts = line.Split(';');
            if (parts.Length < 9) continue;
            string title = parts[1];
            string genre = parts[2];
            string director = parts[3];
            int year = int.Parse(parts[4]);
            bool watched = parts[5] == "Да";
            int rating = int.Parse(parts[6]);
            string notes = parts[7];
            try
            {
                AddMovie(title, genre, director, year, watched, rating, notes);
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Ошибка импорта строки: {ex.Message}");
            }
        }
    }
}

public static class Program
{
    private static string ReadString(string prompt)
    {
        Console.Write(prompt);
        return Console.ReadLine()?.Trim() ?? "";
    }

    private static int ReadInt(string prompt)
    {
        while (true)
        {
            Console.Write(prompt);
            if (int.TryParse(Console.ReadLine(), out int result))
                return result;
            Console.WriteLine("Введите число.");
        }
    }

    private static bool ReadBool(string prompt)
    {
        while (true)
        {
            string input = ReadString(prompt);
            if (input == "1") return true;
            if (input == "0") return false;
            Console.WriteLine("Введите 1 или 0.");
        }
    }

    private static void PrintMovie(Movie movie)
    {
        string status = movie.Watched ? "✅ Просмотрен" : "⏳ Не просмотрен";
        Console.WriteLine($"#{movie.Id} - {movie.Title} ({movie.Year})");
        Console.WriteLine($"   Жанр: {movie.Genre}, Режиссёр: {movie.Director}");
        Console.WriteLine($"   {status}, Оценка: {movie.Rating}/10");
        if (!string.IsNullOrWhiteSpace(movie.Notes))
            Console.WriteLine($"   Заметки: {movie.Notes}");
        Console.WriteLine($"   Добавлен: {movie.AddedDate}");
    }

    public static void Main()
    {
        var randomizer = new MovieRandomizer();
        try { randomizer.LoadFromFile("movies_data.json"); }
        catch { Console.WriteLine("Не удалось загрузить данные."); }

        while (true)
        {
            Console.WriteLine("\n===== ГЕНЕРАТОР СЛУЧАЙНОГО ФИЛЬМА (C#) =====");
            Console.WriteLine("1. Добавить фильм");
            Console.WriteLine("2. Показать все фильмы");
            Console.WriteLine("3. Рекомендовать случайный фильм");
            Console.WriteLine("4. Рекомендовать по жанру");
            Console.WriteLine("5. Показать непросмотренные фильмы");
            Console.WriteLine("6. Отметить фильм как просмотренный");
            Console.WriteLine("7. Редактировать фильм");
            Console.WriteLine("8. Удалить фильм");
            Console.WriteLine("9. Показать статистику");
            Console.WriteLine("10. Сохранить в файл");
            Console.WriteLine("11. Загрузить из файла");
            Console.WriteLine("12. Экспорт в CSV");
            Console.WriteLine("13. Импорт из CSV");
            Console.WriteLine("0. Выход");
            string choice = ReadString("Выберите действие: ");

            switch (choice)
            {
                case "0": return;
                case "1":
                    string title = ReadString("Название: ");
                    if (string.IsNullOrWhiteSpace(title)) { Console.WriteLine("Название не может быть пустым."); continue; }
                    string genre = ReadString("Жанр: ");
                    if (string.IsNullOrWhiteSpace(genre)) { Console.WriteLine("Жанр не может быть пустым."); continue; }
                    string director = ReadString("Режиссёр (необязательно): ");
                    int year = ReadInt("Год выпуска: ");
                    bool watched = ReadBool("Статус (1-просмотрен, 0-нет): ");
                    int rating = ReadInt("Оценка (1-10): ");
                    string notes = ReadString("Заметки (необязательно): ");
                    try
                    {
                        var movie = randomizer.AddMovie(title, genre, director, year, watched, rating, notes);
                        Console.WriteLine($"Фильм добавлен с ID {movie.Id}");
                    }
                    catch (Exception ex) { Console.WriteLine($"Ошибка: {ex.Message}"); }
                    break;
                case "2":
                    if (!randomizer.Movies.Any()) Console.WriteLine("Нет фильмов.");
                    else foreach (var m in randomizer.Movies) PrintMovie(m);
                    break;
                case "3":
                    var movie = randomizer.RandomMovie();
                    if (movie == null) Console.WriteLine("Нет фильмов в коллекции.");
                    else { Console.WriteLine("\n🎬 Рекомендую посмотреть:"); PrintMovie(movie); }
                    break;
                case "4":
                    string g = ReadString("Введите жанр: ");
                    if (string.IsNullOrWhiteSpace(g)) { Console.WriteLine("Введите жанр."); continue; }
                    var byGenre = randomizer.RandomMovie(g);
                    if (byGenre == null) Console.WriteLine($"Нет фильмов в жанре '{g}'.");
                    else { Console.WriteLine($"\n🎬 Рекомендую посмотреть в жанре {g}:"); PrintMovie(byGenre); }
                    break;
                case "5":
                    var unwatched = randomizer.GetUnwatched();
                    if (!unwatched.Any()) Console.WriteLine("Нет непросмотренных фильмов.");
                    else foreach (var m in unwatched) PrintMovie(m);
                    break;
                case "6":
                    int id = ReadInt("Введите ID фильма: ");
                    var found = randomizer.FindMovie(id);
                    if (found == null) { Console.WriteLine("Фильм не найден."); continue; }
                    if (found.Watched) Console.WriteLine("Фильм уже отмечен как просмотренный.");
                    else { randomizer.EditMovie(id, new Dictionary<string, object> { ["watched"] = true }); Console.WriteLine("Фильм отмечен как просмотренный."); }
                    break;
                case "7":
                    int eid = ReadInt("Введите ID фильма для редактирования: ");
                    var old = randomizer.FindMovie(eid);
                    if (old == null) { Console.WriteLine("Фильм не найден."); continue; }
                    Console.WriteLine("Оставьте поле пустым, чтобы не менять.");
                    string newTitle = ReadString($"Название ({old.Title}): ");
                    string newGenre = ReadString($"Жанр ({old.Genre}): ");
                    string newDirector = ReadString($"Режиссёр ({old.Director}): ");
                    string newYearStr = ReadString($"Год ({old.Year}): ");
                    string newWatchedStr = ReadString($"Статус (1-просмотрен, 0-нет) сейчас: {(old.Watched ? "1" : "0")}: ");
                    string newRatingStr = ReadString($"Оценка ({old.Rating}): ");
                    string newNotes = ReadString($"Заметки ({old.Notes}): ");
                    var updates = new Dictionary<string, object>();
                    if (!string.IsNullOrWhiteSpace(newTitle)) updates["title"] = newTitle;
                    if (!string.IsNullOrWhiteSpace(newGenre)) updates["genre"] = newGenre;
                    if (!string.IsNullOrWhiteSpace(newDirector)) updates["director"] = newDirector;
                    if (!string.IsNullOrWhiteSpace(newYearStr))
                    {
                        if (int.TryParse(newYearStr, out int y)) updates["year"] = y;
                        else Console.WriteLine("Год должен быть числом, пропускаем.");
                    }
                    if (!string.IsNullOrWhiteSpace(newWatchedStr)) updates["watched"] = newWatchedStr == "1";
                    if (!string.IsNullOrWhiteSpace(newRatingStr))
                    {
                        if (int.TryParse(newRatingStr, out int r)) updates["rating"] = r;
                        else Console.WriteLine("Оценка должна быть числом, пропускаем.");
                    }
                    if (!string.IsNullOrWhiteSpace(newNotes)) updates["notes"] = newNotes;
                    if (randomizer.EditMovie(eid, updates)) Console.WriteLine("Фильм обновлён.");
                    else Console.WriteLine("Ошибка обновления.");
                    break;
                case "8":
                    int delId = ReadInt("Введите ID фильма для удаления: ");
                    if (randomizer.DeleteMovie(delId)) Console.WriteLine("Фильм удалён.");
                    else Console.WriteLine("Фильм не найден.");
                    break;
                case "9":
                    var stats = randomizer.GetStats();
                    Console.WriteLine("\n=== СТАТИСТИКА ===");
                    Console.WriteLine($"Всего фильмов: {stats["total"]}");
                    Console.WriteLine($"Просмотрено: {stats["watched"]}");
                    Console.WriteLine($"Не просмотрено: {stats["unwatched"]}");
                    Console.WriteLine($"Средняя оценка (просмотренные): {stats["avg_rating"]:F2}");
                    Console.WriteLine("По жанрам:");
                    var genres = (Dictionary<string, int>)stats["genres"];
                    foreach (var kv in genres) Console.WriteLine($"  {kv.Key}: {kv.Value}");
                    break;
                case "10":
                    try { randomizer.SaveToFile("movies_data.json"); Console.WriteLine("Сохранено."); }
                    catch (Exception ex) { Console.WriteLine($"Ошибка: {ex.Message}"); }
                    break;
                case "11":
                    try { randomizer.LoadFromFile("movies_data.json"); Console.WriteLine("Загружено."); }
                    catch (Exception ex) { Console.WriteLine($"Ошибка: {ex.Message}"); }
                    break;
                case "12":
                    try { randomizer.ExportCSV("movies_export.csv"); Console.WriteLine("Экспортировано в movies_export.csv"); }
                    catch (Exception ex) { Console.WriteLine($"Ошибка: {ex.Message}"); }
                    break;
                case "13":
                    try { randomizer.ImportCSV("movies_export.csv"); Console.WriteLine("Импортировано из movies_export.csv"); }
                    catch (Exception ex) { Console.WriteLine($"Ошибка: {ex.Message}"); }
                    break;
                default: Console.WriteLine("Неизвестная команда."); break;
            }
        }
    }
}
